prov/util/src/util_mem_monitor.c

```diff
@@ -529,6 +536,99 @@ static void *ofi_uffd_handler(void *arg)
 	return NULL;
 }
 
+static void ofi_uffd_pagefault_handler(struct uffd_msg *msg)
+{
+	struct uffdio_zeropage zp;
+	int i;
+	int ret;
+	void * const address = (void *) (uintptr_t) msg->arg.pagefault.address;
+	uint64_t const flags = (uint64_t) msg->arg.pagefault.flags;
+
+	/* ofi_uffd_register sets the mode to
+	 * UFFDIO_REGISTER_MODE_MISSING.  As a result, we can 
+	 * get read, write or write-protect notifications via
+	 * UFFD_EVENT_PAGEFAULT.  The only ones we can sensibly
+	 * handle are writes to non-backed pages.
+	 * (Read and write-protect nofications are likely
+	 * application bugs.) 
+	 */
+
+	if (UFFD_PAGEFAULT_FLAG_WRITE != flags) {
+		FI_DBG(&core_prov, FI_LOG_MR, 
+		       "UFFD pagefault with unrecognized flags: %lu, address %p\n",
+		       flags, address);
+		return;
+	}
+
+	/* The event tells us the address of the fault
+	 * (which can be anywhere on the page). It does not
+	 * tell us the size of the page so we have to guess
+	 * from the list of known page_sizes.
+	 *
+	 * We employ the standard resolution: install a zeroed page.
+	 */
+
+	for (i = 0; i < num_page_sizes; ) {
+		/* setup a zeropage reqest for this pagesize */
+		zp.range.start = (uint64_t) (uintptr_t)
+			ofi_get_page_start(address, page_sizes[i]);
+		zp.range.len = (uint64_t) page_sizes[i];
+		zp.mode = 0;
+		zp.zeropage = 0;
+
+		ret = ioctl(uffd.fd, UFFDIO_ZEROPAGE, &zp);
+
+		if (0 == ret)		/* success */
+			return;
+
+		/* Note: the documentation (man ioctl_userfaultfd) says
+		 * that the ioctl() returns -1 on error and errno is set
+		 * to indicate the error. It also says that the zeropage
+		 * member of struct uffdio_zeropage is set to the negated
+		 * error.  The unit tests for uffd say 
+		 *    real retval in uffdio_zeropage.zeropage
+		 * so that's what we use here.
+		 */
+
+		if (-EAGAIN == zp.zeropage)
+			/* This is a tough case.  If the memory map is 
+			 * changing, the kernel returns EAGAIN before
+			 * installing the zeroed page.  So the page
+			 * fault has not been rectified.  If we don't try
+			 * again, the application will crash.  If we add 
+			 * a maximum retry count we could still end up
+			 * with an unresolved page fault.
+			 *
+			 * It's likely a kernel bug or (something else
+			 * bad like OOM) if it returns EAGAIN forever.
+			 * So we retry until we get something besides
+			 * EAGAIN.
+			 */ 
+			continue;	/* retry this page size */
+		
+		i++;			/* try next page size */
+
+		if (-EINVAL == zp.zeropage)     /* wrong page size */
+			continue;
```

Comment:
Having a continue here is weird, since there's basically no code below worth skipping.

Reply:
Since we are trying various page sizes and only needs one success, incorrect page size should not be reported as error, so it makes sense to skip the error message.